packages/@n8n/nodes-langchain/nodes/tools/ToolHttpRequest/utils.ts

```diff
@@ -0,0 +1,357 @@
+				}
+				options.qs[queryAuth.name as string] = queryAuth.value;
+				return await ctx.helpers.httpRequest(options);
+			};
+		} else if (genericType === 'httpCustomAuth') {
+			const customAuth = await ctx.getCredentials('httpCustomAuth', itemIndex);
+
+			return async (options: IHttpRequestOptions) => {
+				const auth = jsonParse<IRequestOptionsSimplified>((customAuth.json as string) || '{}', {
+					errorMessage: 'Invalid Custom Auth JSON',
+				});
+				if (auth.headers) {
+					options.headers = { ...options.headers, ...auth.headers };
+				}
+				if (auth.body) {
+					options.body = { ...(options.body as IDataObject), ...auth.body };
+				}
+				if (auth.qs) {
+					options.qs = { ...options.qs, ...auth.qs };
+				}
+				return await ctx.helpers.httpRequest(options);
+			};
+		} else if (genericType === 'oAuth1Api') {
+			return async (options: IHttpRequestOptions) => {
+				return await ctx.helpers.requestOAuth1.call(ctx, 'oAuth1Api', options);
+			};
+		} else if (genericType === 'oAuth2Api') {
+			return async (options: IHttpRequestOptions) => {
+				return await ctx.helpers.requestOAuth2.call(ctx, 'oAuth1Api', options, {
+					tokenType: 'Bearer',
+				});
+			};
+		}
+	} else if (credentialsType === 'predefinedCredentialType') {
+		const predefinedType = ctx.getNodeParameter('nodeCredentialType', itemIndex) as string;
+		const additionalOptions = getOAuth2AdditionalParameters(predefinedType);
+
+		return async (options: IHttpRequestOptions) => {
+			return await ctx.helpers.requestWithAuthentication.call(
+				ctx,
+				predefinedType,
+				options,
+				additionalOptions && { oauth2: additionalOptions },
+				itemIndex,
+			);
+		};
+	}
+
+	return async (options: IHttpRequestOptions) => {
+		return await ctx.helpers.httpRequest(options);
+	};
+};
+
+export const configureResponseOptimizer = (ctx: IExecuteFunctions, itemIndex: number) => {
+	const optimizeResponse = ctx.getNodeParameter('optimizeResponse', itemIndex, false) as boolean;
+
+	if (optimizeResponse) {
+		const responseType = ctx.getNodeParameter('responseType', itemIndex) as
+			| 'json'
+			| 'text'
+			| 'html';
+
+		if (responseType === 'html') {
+			const cssSelector = ctx.getNodeParameter('cssSelector', itemIndex, '') as string;
+			const onlyContent = ctx.getNodeParameter('onlyContent', itemIndex, false) as boolean;
+			let elementsToOmit: string[] = [];
+
+			if (onlyContent) {
+				const elementsToOmitUi = ctx.getNodeParameter('elementsToOmit', itemIndex, '') as
+					| string
+					| string[];
+
+				if (typeof elementsToOmitUi === 'string') {
+					elementsToOmit = elementsToOmitUi.split(',').map((s) => s.trim());
+				}
+			}
+
+			return <T>(response: T) => {
+				if (typeof response !== 'string') {
+					throw new NodeOperationError(
+						ctx.getNode(),
+						`The response type must be a string. Received: ${typeof response}`,
+						{ itemIndex },
+					);
+				}
+				const returnData: string[] = [];
+				const html = cheerio.load(response);
```

Comment:
We could use `@mozilla/readability` and `jsdom` here to cleanly extract the content that's likely relevant to an end-user.

Something like this perhaps:
```typescript
import { JSDOM } from 'jsdom'
import { Readability } from '@mozilla/readability'

const dom = await JSDOM.fromURL(url)
const article = new Readability(dom.window.document, {
    keepClasses: true,
}).parse()
```
and then use `article.content`.

we could also consider using [turndown](https://www.npmjs.com/package/turndown) to convert the html into markdown, which LLM tend to handle better than html IMO.
```typescript
import Turndown from 'turndown'
const markdown = turndown.turndown(article.content)
```

Reply:
@netroy 
what would be advantages over `html-to-text` + `Cheerio`? since we already using such setup for `Html` node

Reply:
`cheerio` is great to either use css selectors to extract text, but leaves the burden of determining the semantics in the markup to the end user.

First download an article via `curl https://www.bbc.com/news/articles/cldd6x6gglxo > news.html`. 

Then try this with `cheerio`:
```javascript
const fs = require('fs');
const cheerio = require('cheerio');
const html = fs.readFileSync('news.html', 'utf8');
const $ = cheerio.load(html);
console.log($('body').text());
```
I got this 
![image](https://github.com/n8n-io/n8n/assets/196144/dabb05ea-4d23-4eb3-b8bd-3d4ab8d3d92f)


With `readability` + `turndown`:
```javascript
(async () => {
	const { JSDOM } = require('jsdom');
	const { Readability } = require('@mozilla/readability');
	const Turndown = require('turndown');

	const dom = await JSDOM.fromFile('news.html');
	const article = new Readability(dom.window.document, {
		keepClasses: true,
	}).parse();
	const turndown = new Turndown({
		headingStyle: 'atx',
		hr: '---',
		bulletListMarker: '-',
		codeBlockStyle: 'fenced',
	});
	const markdown = turndown.turndown(article.content);
	console.log(markdown);
})();
```
I got this
![image](https://github.com/n8n-io/n8n/assets/196144/9319fa1f-e70b-42ac-a39e-412725790eac)


Perhaps we should add a "Extract as Markdown" option in the node to determine if we want to use markup to reduce semantic noise in the extracted text?

Reply:
+1 for using `@mozilla/readability`